setSuggestedFlow api improvements

knolleary opened the pull request:
Some improvements to the `setSuggestedFlow` api:

### Allow an existing node to be wired to the suggestion

You can now specify an existing 'source' node (and port number), which will be wired to the suggested flow.

```
{
   nodes: [ ... ],
   source: <sourceNode>,
   sourcePort: <portIndex (0-indexed)>
}
```

For example, all of the ghost nodes here were passed in via the `nodes` array. The existing Inject node was passed in as `source` and `sourcePort` set to `0` (the default).

<img width="1020" alt="image" src="https://github.com/user-attachments/assets/539b64bd-ef50-41e4-99a0-e9d4ec9bd972" />

### Enable click-to-apply on the ghost nodes

Setting `clickToApply: true` on the provided suggestion will enable event handles on the ghost nodes - clicking on them will automatically apply the flow.

```
{
   nodes: [ ... ],
   clickToApply: true
}
```

Steve-Mcl commented:
Nick there are some niggles with this:
1. Placement of a node when dropped to the left, the wire (or node) placement is not aligned
    ![chrome_d4d55CTY4b](https://github.com/user-attachments/assets/1e8a45f6-6069-4d4f-aef4-19bf9ab11808)
2. Should the `esc` key and/or _click elsewhere_ cause the suggestion to vanish?
    ![chrome_2XfFt5n8J8](https://github.com/user-attachments/assets/200ea4dc-83df-44e7-934a-9a7737225e79)

Question:
* Can you think of a way we could accept the suggestion via keyboard? (tab or right arrow?)
* Can you think of a way we could change suggestion (we are working on the idea that top 3 suggestions will be available) - e.g. down arrow permits alternate suggestion to be set?

Please note, the up/down to select alternative would be moot if we proceed as discussed in recent meeting - e.g. show suggestions in the typeSearch (then user can see and select alternatives) - that leads nicely on to the points I raised in my comment here: https://github.com/node-red/node-red/pull/5135#issuecomment-2993634356

knolleary commented:
Have now added a plugin api to allow custom providers of suggested nodes to the quick add TypeSearch Dialog.

```
RED.plugins.registerPlugin('demo-suggestion-source', {
            type: 'node-red-flow-suggestion-source',
            getSuggestions: async function (context) {
                const suggestion = {
                    label: 'Change/Debug Combo',
                    nodes: [
                        { id: 'suggestion-1', type: 'change', x: 0, y: 0, wires:[['suggestion-2']] },
                        { id: 'suggestion-2', type: 'function', outputs: 3, x: 200, y: 0, wires:[['suggestion-3'],['suggestion-4'],['suggestion-6']] },
                        { id: 'suggestion-3', _g: 'suggestion-group-1', type: 'debug', x: 375, y: -40 },
                        { id: 'suggestion-4', _g: 'suggestion-group-1', type: 'debug', x: 375, y: 0 },
                        { id: 'suggestion-5', _g: 'suggestion-group-1', type: 'debug', x: 410, y: 40 },
                        { id: 'suggestion-6', type: 'junction', wires: [['suggestion-5']], x:325, y:40 }
                    ]
                }
                const suggestion2 = {
                    label: 'Another Change/Debug Combo',
                    nodes: [
                        { id: 'suggestion-1', type: 'change', x: 0, y: 0, wires:[['suggestion-2']] },
                        { id: 'suggestion-2', type: 'function', outputs: 3, x: 200, y: 0 },
                    ]
                }
                await new Promise(resolve => setTimeout(resolve, 1000)); // Simulate async operation
                return [suggestion, suggestion2]
            }
        })
```

 - `type` must be `node-red-flow-suggestion-source`
 - plugin must provide a `getSessions` async function. This will get called when the TypeSearch is opened.
 - The function receives a `context` object - with details of where the TypeSearch is being triggered. It is an object with the properties:
  - `workspace`
  - `source` - the node being wired to
  - `sourcePort` - 0-index port numnber
  - `sourcePortType` - 0: output 1:input (dbl check I got that the right way around)

knolleary commented:
If the suggestion includes a `source` node, it can now also specify `position: 'relative'` and the suggested nodes will be placed relative to the source node.

The snapping to grid isn't quite right yet, but this is close enough to unblock progress.

Steve-Mcl commented:
Morning Nick. Now I have got plugin (mostly) written it has revealed some artefacts. 

It would take a wall of text to explain them clearly so I prepared gifs with TLDR to (hopefully) aid comprehension.


#### Undo leaves ghost
![chrome_E2bI6YX8s4](https://github.com/user-attachments/assets/326c509d-52a2-482b-b382-3602b44bb87b)


#### Deleted accepted suggestion leaves dangling wire
![chrome_P9WJOteTiX](https://github.com/user-attachments/assets/f5bff8b5-a86f-424e-9dc4-a23292c285fb)


#### Ignoring suggestion and undoing prev leaves ghost with dangling wire
![chrome_E2bI6YX8s4](https://github.com/user-attachments/assets/05ba5c61-3a33-44f6-8297-b8ca0599c7d8)


---

I suspect these are all somewhat related but I think they reveal some behaviours we should support.
1. Clicking anything but the ghost should dismiss the ghost 
2. Any key press other than `TAB` should dismiss the ghost 

I think that ^ might covert it all?

knolleary commented:
Updates pushed:

1. removed actions and hardcoded keyboard handling to be `tab` to apply and *anything else* to clear
2. updated click handling so a click on the node applies and a click *anywhere else* clears

Looking at the left-behind wire on undo next.

knolleary commented:
Fixes pushed:

 - NLS the 'loading suggestions' message
 - Don't require an x/y when providing a single node to the api
 - Avoid redrawing the same suggestion multiple times when triggered by mouse-enter events in typeSearch (ie, don't reapply the same suggestion that is already applied)
